tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java

```diff
@@ -421,65 +444,79 @@ public TaskSpec call() throws Exception {
-            if (task.state == ManagedTaskState.RUNNING) {
-                task.state = ManagedTaskState.STOPPING;
-                broadcastStop = true;
-            }
-            task.maybeSetError(error);
-            task.activeWorkers.remove(nodeName);
-            if (task.activeWorkers.size() == 0) {
-                task.doneMs = time.milliseconds();
-                task.state = ManagedTaskState.DONE;
-                log.info("Task {} is now complete on {} with error: {}", id,
-                    Utils.join(task.workers, ", "),
-                    task.error.isEmpty() ? "(none)" : task.error);
-            } else if (broadcastStop) {
-                log.info("Node {} stopped.  Stopping task {} on worker(s): {}",
-                    id, Utils.join(task.activeWorkers, ", "));
-                for (String workerName : task.activeWorkers) {
-                    nodeManagers.get(workerName).stopWorker(id);
-                }
+            WorkerState prevState = task.workerStates.get(nodeName);
+            log.debug("Task {}: Updating worker state for {} from {} to {}.",
+                id, nodeName, prevState, state);
+            task.workerStates.put(nodeName, state);
+            if (state.done() && (!prevState.done())) {
+                handleWorkerCompletion(task, nodeName, (WorkerDone) state);
             }
             return null;
         }
     }
 
+    /**
+     * Handle a worker being completed.
+     *
+     * @param task      The task that owns the worker.
+     * @param nodeName  The name of the node on which the worker is running.
+     * @param state     The worker state.
+     */
+    private void handleWorkerCompletion(ManagedTask task, String nodeName, WorkerDone state) {
+        if (state.error().isEmpty()) {
+            log.info("{}: Worker {} finished with status '{}'",
+                nodeName, task.id, JsonUtil.toJsonString(state.status()));
+        } else {
+            log.warn("{}: Worker {} finished with error '{}' and status '{}'",
+                nodeName, task.id, state.error(), JsonUtil.toJsonString(state.status()));
+            task.maybeSetError(state.error());
+        }
+        if (task.activeWorkers().isEmpty()) {
+            task.doneMs = time.milliseconds();
+            task.state = ManagedTaskState.DONE;
+            log.info("Task {} is now complete on {} with error: {}", task.id,
+                Utils.join(task.workerStates.keySet(), ", "),
+                task.error.isEmpty() ? "(none)" : task.error);
+        } else if (task.state == ManagedTaskState.RUNNING) {
+            TreeSet<String> activeWorkers = task.activeWorkers();
+            log.info("Node {} stopped.  Stopping task {} on worker(s): {}",
+                task.id, Utils.join(activeWorkers, ", "));
+            task.state = ManagedTaskState.STOPPING;
+            for (String workerName : activeWorkers) {
+                nodeManagers.get(workerName).stopWorker(task.id);
+            }
+        }
```

Comment:
It looks like before moving this code into a separate method, we handled PENDING ManagedTaskState. Or this state is not possible when we are in this method? Maybe we should check and throw an exception in that case? Also, what if the task is in STOPPING state?